### PR TITLE
fix(engine): make the drive the sole PlanetScale cutover actor

### PR DESCRIPTION
**Pull request by aparajon**
## Why this matters

Deploy requests were created with PlanetScale's `auto_cutover` enabled whenever cutover wasn't deferred, so two actors raced to complete every cutover: PlanetScale's own auto-cutover machinery and the drive's `ApplyDeployRequest` call at `pending_cutover`. That race produced spurious "changes have not been staged" rejections, and worse, meant a schema could cut over with no SchemaBot involvement or caller attribution if the drive lost the race.

## What it does

- Creates deploy requests with `auto_cutover` disabled, always. The deploy request parks at `pending_cutover` and exactly one actor completes it: the drive (or an operator, when cutover is deferred).
- Drops the now-constant parameter from `createDeployRequest` and the resume path's unused defer-cutover read.
- Adds a payload test asserting the engine never enables PlanetScale's auto-cutover.
- Bounds the drive's auto-cutover retry: a permanent rejection fails the apply immediately, and repeated consecutive hard rejections settle it (failed, or paused for operator retry) instead of retrying forever — mirroring the progress-poll bound. Each hard rejection increments a new `schemabot.apply.auto_cutover_failure.total` counter. Not-ready rejections stay self-clearing and reset the counter.
- Fixes the progress comment for non-deferred applies parked at `waiting_for_cutover`: the footer now says cutover triggers automatically instead of offering the manual `schemabot cutover` command; deferred applies keep the command.
- Sweeps the docs that still said PlanetScale cuts over automatically (engine/state READMEs, control edge-case catalog, engine crash-resume comment) to state the drive-as-sole-cutover-actor invariant.

```
before:  copy done → PS auto-cutover ──┐
                     drive Cutover() ──┴─ race, double-fire
after:   copy done → pending_cutover → drive Cutover() (sole actor)
```

A deploy request now only moves past `pending_cutover` when SchemaBot acts, so cutover timing, retries, and attribution are fully owned by the drive.

**By design, this removes autonomous completion.** Previously, a non-deferred apply whose driver died and was never reclaimed could still complete via PlanetScale's auto-cutover. Now it parks at `pending_cutover` until a drive claims it — an unattended apply stays incomplete rather than cutting over with no owner, which is the intended trade: no schema change lands without SchemaBot acting.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/engine/planetscale/branch.go [hunk 1]
```diff
@@ -500,13 +500,18 @@ func (e *Engine) waitForBranchReady(ctx context.Context, client psclient.PSClien
 	}
 }
 
-func (e *Engine) createDeployRequest(ctx context.Context, client psclient.PSClient, org, database, branchName, intoBranch string, autoCutover, autoDeleteBranch bool) (*ps.DeployRequest, error) {
+func (e *Engine) createDeployRequest(ctx context.Context, client psclient.PSClient, org, database, branchName, intoBranch string, autoDeleteBranch bool) (*ps.DeployRequest, error) {
+	// AutoCutover stays off so SchemaBot is the sole cutover actor: the deploy
+	// request parks at pending_cutover and the drive completes it via Cutover
+	// (or an operator does, when cutover is deferred). Letting PlanetScale cut
+	// over on its own would race the drive's cutover call and move the schema
+	// without SchemaBot's involvement or caller attribution.
```
[the remark this answers is not in the document]
Comment (aparajon) [reply pkg/engine/planetscale/branch.go:508]: 🤖 Fixed in 20ff97cf — the comment now says "the driver completes it via Cutover" / "the driver's cutover call", matching the repo's driver-noun terminology.

*Generated by Claude Code (Claude Fable 5)*

File: pkg/engine/planetscale/branch_test.go [hunk 2]
```diff
@@ -88,3 +92,34 @@ func TestDiffKeyspace_DetectsSchemaChanges(t *testing.T) {
...
+func (c *deployRequestRecorderClient) CreateDeployRequest(_ context.Context, req *ps.CreateDeployRequestRequest) (*ps.DeployRequest, error) {
+	c.created = req
+	return &ps.DeployRequest{Number: 1}, nil
+}
+
+// SchemaBot is the sole cutover actor: deploy requests are created with
+// PlanetScale's auto-cutover disabled so they park at pending_cutover until
+// the drive (or a deferred-cutover operator) completes the cutover. If
+// PlanetScale cut over on its own, the schema could move without SchemaBot's
+// involvement or caller attribution.
```
[the remark this answers is not in the document]
Comment (aparajon) [reply pkg/engine/planetscale/branch_test.go:112]: 🤖 Fixed in 20ff97cf — the test comment now uses "the driver (or a deferred-cutover operator) completes the cutover".

*Generated by Claude Code (Claude Fable 5)*

File: pkg/engine/planetscale/branch.go [hunk 3]
```diff
@@ -500,13 +500,18 @@ func (e *Engine) waitForBranchReady(ctx context.Context, client psclient.PSClien
 	}
 }
 
-func (e *Engine) createDeployRequest(ctx context.Context, client psclient.PSClient, org, database, branchName, intoBranch string, autoCutover, autoDeleteBranch bool) (*ps.DeployRequest, error) {
+func (e *Engine) createDeployRequest(ctx context.Context, client psclient.PSClient, org, database, branchName, intoBranch string, autoDeleteBranch bool) (*ps.DeployRequest, error) {
```
[the remark this answers is not in the document]
Comment (aparajon) [reply pkg/engine/planetscale/branch.go:503]: 🤖 Not applying this one: every deploy-request helper in this file (createDeployRequest, getDeployRequest, and the control-path mutations) takes psclient.PSClient, and the package's established test idiom is a fake that embeds psclient.PSClient and overrides the one method under test (see control_test.go). Introducing a one-off single-method interface for this helper would make it inconsistent with its siblings without changing any behavior.

*Generated by Claude Code (Claude Fable 5)*

**Comment by Kiran01bm** [comment]
🤖 Review findings - created by Kiran's code review agent - for pull/763.

  **Verdict:** 2 findings — the code change is correct and the drive-side
  cutover machinery holds up under attack, but the PR removes PlanetScale's
  autonomous completion backstop without bounding the newly load-bearing retry
  loop, and flips a documented invariant in five doc locations without
  sweeping them.

  ### Findings

  **1. [PLAUSIBLE] The non-deferred `waiting_for_cutover` branch — newly
  load-bearing — has no timeout, no metric, and no failure transition when the
  drive's cutover fails persistently.**
  Pre-PR, PlanetScale's auto-cutover was an independent fallback actor: even
  if the drive's `ApplyDeployRequest` calls failed, PlanetScale completed the
  deploy. Post-PR the drive is the sole actor, and its auto-trigger
  ([pkg/tern/local_apply_grouped.go#L782-L790](https://github.com/block/schema
  bot/blob/20ff97cf737f8eea919fae58c679767df649a890/pkg/tern/local_apply_group
  ed.go#L782-L790)) handles failure by logging and retrying next tick:

  ```go
  if _, err := eng.Cutover(ctx, controlReq); err != nil {
      c.logger.Error("auto-cutover failed", append(apply.LogAttrs(), "error",
  err)...)
  }
  ```

  The `waitingForManualActionTimeout` (14d) cancel is gated on
  `opts.DeferCutover`
  ([#L808-L809](https://github.com/block/schemabot/blob/20ff97cf737f8eea919fae
  58c679767df649a890/pkg/tern/local_apply_grouped.go#L808-L809)), the
  consecutive-error abort counts only `Progress` failures, and no metric is
  emitted from this path.

  Verified narrowing (why PLAUSIBLE, not CONFIRMED): most trigger scenarios
  self-terminate — an out-of-band deleted deploy request surfaces as a
  `PermanentError` on the next progress poll
  ([progress.go#L56-L60](https://github.com/block/schemabot/blob/20ff97cf737f8
  eea919fae58c679767df649a890/pkg/engine/planetscale/progress.go#L56-L60)),
  and full token loss kills progress polling too (10-consecutive-error abort).
  The surviving mode is mutation-only failure (cutover/apply scope revoked
  while reads still work, or a persistent PS-side 4xx at `pending_cutover`):
  the apply then pins at `waiting_for_cutover` indefinitely, hitting the PS
  API and writing an "Auto-triggering cutover" apply-event **twice per second,
  forever**, holding the database's deploy queue, lock, and check-run gate
  until a human cancels. Suggested hardening: a consecutive-cutover-failure
  counter (mirroring the progress one) that fails the apply, plus a metric —
  this also satisfies the repo's "metrics should be actionable" bar for
  dangerous branches.

  **2. [CONFIRMED] The PR flips a documented invariant — "PlanetScale cuts
  over automatically for non-deferred applies" — in five doc locations it
  doesn't sweep, including the operator-facing incident doc.**
  All verified wrong at the PR head:
  - [docs/grpc-control-edge-cases.md#L149-L151](https://github.com/block/schem
  abot/blob/20ff97cf737f8eea919fae58c679767df649a890/docs/grpc-control-edge-ca
  ses.md#L149-L151): "Vitess/PlanetScale creates the deploy request with
  auto-cutover enabled" (also the scenario table at
  [#L181](https://github.com/block/schemabot/blob/20ff97cf737f8eea919fae58c679
  767df649a890/docs/grpc-control-edge-cases.md#L181) — "Progress advances
  through deploy/cutover to completed" — and the review checklist at
  [#L369](https://github.com/block/schemabot/blob/20ff97cf737f8eea919fae58c679
  767df649a890/docs/grpc-control-edge-cases.md#L369)). An operator triaging a
  non-deferred apply parked at `pending_cutover` with a dead driver would
  conclude PlanetScale will finish on its own and wait — the opposite of the
  required action (ensure a driver reclaims and cuts over).
  - [pkg/engine/planetscale/planetscale.go#L139-L141](https://github.com/block
  /schemabot/blob/20ff97cf737f8eea919fae58c679767df649a890/pkg/engine/planetsc
  ale/planetscale.go#L139-L141): "If SchemaBot crashes, the deploy
  continues... no special resume logic needed beyond polling" — post-PR,
  polling alone never completes a crashed non-deferred apply; the resumed
  drive must issue `Cutover`.
  - [pkg/state/README.md#L77](https://github.com/block/schemabot/blob/20ff97cf
  737f8eea919fae58c679767df649a890/pkg/state/README.md#L77):
  "`waiting_for_cutover`/`cutting_over`: Only with `--defer-cutover` or atomic
  mode (Spirit)" — every non-instant PlanetScale apply now passes through
  both; also [#L214](https://github.com/block/schemabot/blob/20ff97cf737f8eea9
  19fae58c679767df649a890/pkg/state/README.md#L214) ("PlanetScale's deploy
  request triggers this automatically when all shards are ready") and the
  state table at [#L28](https://github.com/block/schemabot/blob/20ff97cf737f8e
  ea919fae58c679767df649a890/pkg/state/README.md#L28) ("waiting for manual
  cutover (atomic mode only)").
  - [pkg/engine/README.md#L57-L76](https://github.com/block/schemabot/blob/20f
  f97cf737f8eea919fae58c679767df649a890/pkg/engine/README.md#L57-L76): the
  state diagram's "(auto) → CuttingOver" direct edge no longer exists for
  PlanetScale; `WaitingForCutover` no longer implies `defer_cutover=true`.
  Code written against this interface doc (alerting/UX keyed on
  `waiting_for_cutover` meaning "operator action required") would fire on
  every routine apply.

  ### The one thing that could have broken, verified

  Whether any path that previously relied on PlanetScale's auto-cutover to
  complete now strands at `pending_cutover`. Verified safe on every enumerated
  path: the tick loop auto-triggers `eng.Cutover` for non-deferred applies
  ungated by any control request
  ([local_apply_grouped.go#L783](https://github.com/block/schemabot/blob/20ff9
  7cf737f8eea919fae58c679767df649a890/pkg/tern/local_apply_grouped.go#L783));
  driver death at `pending_cutover` recovers via the claimable
  `waiting_for_cutover` state, a reattach-only resume
  (`deployRequestNeedsResumeDeploy` only re-deploys `ready`+undeployed DRs),
  and the resumed poll loop's same auto-trigger; ordered-cutover fan-out
  drives force `DeferCutover=false`; and the LocalScale fake honors
  `auto_cutover=false` (only auto-completes when the flag is set, 409s
  `apply-deploy` unless the DR is at `pending_cutover`), so the green
  non-deferred Vitess e2e suites genuinely exercised the drive-only cutover
  path — they would have hung if it didn't work.

  ### Notes (not blocking)

  - **Misleading footer during the new deterministic dwell** — [pkg/webhook/te
  mplates/apply.go#L957-L958](https://github.com/block/schemabot/blob/20ff97cf
  737f8eea919fae58c679767df649a890/pkg/webhook/templates/apply.go#L957-L958):
  the "To proceed with cutover: `schemabot cutover …`" footer keys on state
  alone, and a state change bypasses the comment observer's rate limiter, so
  every non-instant non-deferred apply now briefly posts a manual-cutover
  instruction the drive has already acted on. An obeying operator double-fires
  `ApplyDeployRequest` — `requestCutover` has no "not deferred" rejection
  guard ([local_control.go#L26-L36](https://github.com/block/schemabot/blob/20
  ff97cf737f8eea919fae58c679767df649a890/pkg/tern/local_control.go#L26-L36)) —
  yielding an error log, a failed control request, and one drive
  claim-exit/re-claim cycle. Self-healing, no corruption; cheap fix is
  plumbing the observer's existing `shouldDeferCutover` into
  `ApplyStatusCommentData` and gating the footer.
  - **Autonomous completion is gone by design — worth stating in the PR
  body**: a non-deferred apply whose driving process dies and is never
  reclaimed (local/CLI mode with no server operator) previously completed via
  PlanetScale; it now parks at `pending_cutover` holding the database's deploy
  queue until manually resumed or cancelled. Intended trade (attribution over
  autonomy), but it's the one previously-guaranteed completion path that no
  longer exists.
  - **Coverage**: the new payload test pins the helper's request, but a future
  direct `client.CreateDeployRequest` call would bypass it silently, and
  there's no named test for the drive auto-trigger invariant at
  `local_apply_grouped.go:783` (a regression there would surface only as
  opaque e2e timeouts). Also unverifiable from the repo: whether PlanetScale
  database-level deploy settings can override the request-level
  `AutoCutover:false`, and whether real PS reports `pending_cutover` before
  changes are fully staged (which would Error-log "auto-cutover failed …
  changes have not been staged" once per apply on the golden path before the
  next-tick retry succeeds).

  ### Verified correct

  - Both `createDeployRequest` call sites updated with correct positional
  arguments (no bool misalignment — `resumeApply`'s trailing `true` was
  already `autoDeleteBranch`); the helper is the only caller of
  `client.CreateDeployRequest`; `psclient` is a pure passthrough with no
  `AutoCutover` defaulting layer.
  - The surviving `deferCutover` uses in `Apply` (logging, instant-DDL
  messaging) remain coherent; the resume path's deleted `defer_cutover` read
  had no other consumer; `defer_deploy` handling kept.
  - Out-of-band deploy-request deletion, cancellation, or error all
  self-terminate the poll loop (PermanentError / cancelled / failed state
  mappings); full token loss self-terminates via the
  consecutive-progress-error abort.
  - The deferred-cutover recovery path (`shouldInspectDeferredCutoverSignal`)
  is gated on stored options plus a Spirit-only capability interface — a
  non-deferred PlanetScale apply dwelling at `pending_cutover` cannot trigger
  it.
  - The new test compiles and passes; the recorder's nil embedded
  `psclient.PSClient` is never invoked; comments describe the invariant (not
  history), use "driver" terminology, and conform to AGENTS.md test
  conventions.
  - CI fully green (all 33 checks), and green here is strong evidence:
  LocalScale enforces the parked `pending_cutover` behavior, so the e2e suites
  exercised the new sole-actor path for real.

  *This review was generated by Claude Code (claude-fable-5).*

**Review by Kiran01bm — APPROVED** [review]

**Comment by aparajon** [comment]
🤖 Thanks for the review — all of it is addressed on this branch (new commits, no rewrite since the approval):

**Finding 1 (unbounded auto-cutover retry)** — fixed in 81d4d659. Auto-cutover now mirrors the progress-poll consecutive-error bound: a `PermanentError` rejection fails the apply immediately, and 10 consecutive hard rejections settle it (paused for operator retry when the error class is retryable, failed otherwise) instead of holding the deploy queue and lock forever. Each hard rejection increments a new `schemabot.apply.auto_cutover_failure.total` counter. Not-ready rejections stay self-clearing, reset the counter, and remain bounded by the escalation from #762. Covered by two new tests plus an assertion that not-ready rejections never count toward the bound.

**Finding 2 (stale "cuts over automatically" docs)** — swept in c8733bfc: the control edge-case catalog (setup prose, non-deferred scenario table, review checklist), the PlanetScale engine's crash-resume comment (which now also states that a crashed drive leaves the deploy request parked at `pending_cutover`), the state README (state table, lifecycle notes, `postpone_completion` shard section), and the engine README state diagram (the `(auto)` edge is now labeled as engine-internal cutover for MySQL/Spirit, with a note on who issues `Cutover()`).

**Note A (misleading manual-cutover footer)** — fixed in c8733bfc via your suggested route: the apply's durable defer-cutover option is plumbed into `ApplyStatusCommentData`, and the `waiting_for_cutover` footer now tells the operator no action is needed unless cutover is actually deferred. Added a preview for the automatic variant and regenerated TEMPLATES.md.

**Note B** — the PR body now states explicitly that autonomous completion is removed by design: an apply whose driver dies and is never reclaimed parks at `pending_cutover` rather than completing with no owner.

CI is green (32/32) on the new head.

*This reply was generated by Claude Code (claude-fable-5).*
